src/network/netdev/wireguard.c

```diff
@@ -0,0 +1,714 @@
+/***
+    This file is part of systemd.
+
+    Copyright 2016-2017 Jörg Thalheim <joerg@thalheim.io>
+    Copyright 2015-2017 Jason A. Donenfeld <Jason@zx2c4.com>. All Rights Reserved.
+
+    systemd is free software; you can redistribute it and/or modify it
+    under the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation; either version 2.1 of the License, or
+    (at your option) any later version.
+
+    systemd is distributed in the hope that it will be useful, but
+    WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+    Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with systemd; If not, see <http://www.gnu.org/licenses/>.
+***/
+
+#include <sys/ioctl.h>
+#include <net/if.h>
+
+#include "alloc-util.h"
+#include "parse-util.h"
+#include "fd-util.h"
+#include "strv.h"
+#include "hexdecoct.h"
+#include "string-util.h"
+#include "wireguard.h"
+#include "networkd-link.h"
+#include "networkd-util.h"
+#include "networkd-manager.h"
+#include "wireguard-netlink.h"
+
+static void resolve_endpoints(NetDev *netdev);
+
+static WireguardPeer *wireguard_peer_new(Wireguard *w, unsigned section) {
+        WireguardPeer *peer;
+
+        assert(w);
+
+        if (w->last_peer_section == section && w->peers)
+                return w->peers;
+
+        peer = new0(WireguardPeer, 1);
+        peer->flags = WGPEER_F_REPLACE_ALLOWEDIPS;
+        if (!peer)
+                return NULL;
+
+        LIST_PREPEND(peers, w->peers, peer);
+        w->last_peer_section = section;
+
+        return peer;
+}
+
+static int set_wireguard_interface(NetDev *netdev) {
+        int r;
+        unsigned int i, j;
+        WireguardPeer *peer, *peer_start;
+        WireguardIpmask *mask, *mask_start = NULL;
+        _cleanup_free_ sd_netlink_message *message = NULL;
+        Wireguard *w;
+        uint32_t serial;
+
+        assert(netdev);
+        w = WIREGUARD(netdev);
+        assert(w);
+
+        peer_start = w->peers;
+
+        do {
+          free(message);
+          message = NULL;
```

Comment:
indenting is off, please stick to multiples of 8 spaces.

Also:

message = mfree(message)